### PR TITLE
android: Integrate Android JUnit tests into CI

### PR DESCRIPTION
**Pull request by arjungm**
Enable on-host execution of Android JUnit (Robolectric) tests within
the CI pipeline. This allows for automated validation of Android-
specific Java code without requiring physical devices or emulators.

The integration ensures that test results are correctly reported by
converting Chromium test output to standard XML formats. It also
refines the build and archiving logic to package the required JDK and
third-party dependencies. Sparse checkout support is added to the CI
environment to efficiently fetch these prerequisites and fix path
resolution issues. GN for robolectric binary is extended to generate
runtime deps files to ensure the archive is complete.

Bug: 431780245

### CODE REVIEW

File: .github/actions/on_host_tests/action.yaml [hunk 1]
```diff
@@ -117,7 +141,86 @@ runs:
...
+
+
+        # 3. Expose the Git sparse checkout roots to the extracted unit_test/ workspace
+        ln -sf "${GITHUB_WORKSPACE}/testing" unit_test/testing
+        ln -sf "${GITHUB_WORKSPACE}/build" unit_test/build
+        ln -sf "${GITHUB_WORKSPACE}/cobalt" unit_test/cobalt
+        ln -sf "${GITHUB_WORKSPACE}/starboard" unit_test/starboard
+        ln -sf "${GITHUB_WORKSPACE}/.github" unit_test/.github
+        ln -sf "${GITHUB_WORKSPACE}/third_party/catapult" unit_test/third_party/catapult
+        ln -sf "${GITHUB_WORKSPACE}/third_party/logdog" unit_test/third_party/logdog
```
Comment (arjungm) [inline .github/actions/on_host_tests/action.yaml]: this whole section is a bit puzzling because it seems like we're:

1. symlinking to `unit_test/third_party` subdirs from `${GITHUB_WORKSPACE}/third_party`
2. symlinking to `${GITHUB_WORKSPACE}/third_party` subdirs to `unit_test/third_party`

I'm not sure why but it might be to ensure that the relative paths for everything else works? this is very bizarre ...

File: .github/actions/on_host_tests/action.yaml [hunk 2]
```diff
@@ -65,9 +69,22 @@ runs:
         # This also prevents accidental overwriting when collecting results.
         test_output="${results_dir}/shard_${{ matrix.shard }}"
         mkdir -p ${test_output}
+        echo "test_output=${test_output}" >> $GITHUB_ENV
+
```
Comment (oxve) [inline .github/actions/on_host_tests/action.yaml:72]: nit: extra newline

File: .github/actions/on_host_tests/action.yaml [hunk 3]
```diff
@@ -27,34 +31,34 @@ runs:
       uses: actions/download-artifact@v5
       with:
         name: ${{ inputs.test_artifacts_key }}
+    - name: Checkout JUnit Dependencies
+      if: ${{ inputs.test_type == 'junit' }}
+      shell: bash
+      run: |
+        git config --global --add safe.directory "${GITHUB_WORKSPACE:-/github/workspace}"
+        git config --global --add safe.directory "*"
```
Comment (oxve) [inline .github/actions/on_host_tests/action.yaml]: This seems a little aggressive. In [.github/actions/checkout/action.yaml](https://github.com/youtube/cobalt/blob/main/.github/actions/checkout/action.yaml) we only tag `${GITHUB_WORKSPACE}/${DEST_PATH}` which is the checkout path.

File: .github/actions/on_host_tests/action.yaml [hunk 4]
```diff
@@ -27,34 +31,34 @@ runs:
...
+        git sparse-checkout add --skip-checks testing build third_party/catapult third_party/logdog
```
Comment (oxve) [inline .github/actions/on_host_tests/action.yaml:39]: These are not declared deps in GN?

File: .github/actions/on_host_tests/action.yaml [hunk 5]
```diff
@@ -117,7 +135,65 @@ runs:
...
+          fi
+
+          # Run JUnit tests and generate JSON report, then convert to XML.
+          #
+          # NOTE: Even though these are junit tests, due to upstream implementation of the Robolectric tests they are
+          # only able to output JSON results. The rest of our CI pipeline is expecting XML so in the short term we are
+          # taking a shortcut to convert JSON to XML for expediency. Given that the Robolectric JSON output is the how
+          # Chromium infrastructure handles test results in general, we will eventually refactor the gtest results to
+          # also output JSON and then subsequently remove this conversion script after adopting JSON as a universal
+          # test result format in our testing workflow.
```
Comment (oxve) [inline .github/actions/on_host_tests/action.yaml:184]: Could you put this in a bug and add a TODO here?

File: .github/actions/on_host_tests/action.yaml [hunk 6]
```diff
@@ -117,7 +135,65 @@ runs:
...
           fi
         done
-        echo "Finished running tests..."
+
+        echo "Finished running GTest tests..."
+        if [[ -n "${failed_suites}" ]]; then
+          echo "Test suites failed:${failed_suites}"
+          # Fail the job so it's easy to retrigger.
+          exit 1
+        fi
```
Comment (oxve) [inline .github/actions/on_host_tests/action.yaml:142]: nit: Missing separating newline

File: build/android/test_runner.py
Comment (oxve) [inline build/android/test_runner.py:1]: This is a Chromium file. I don't know if we should be changing this here. Is this change necessary for getting the tests running?

File: build/config/android/rules.gni
Comment (oxve) [inline build/config/android/rules.gni:1]: These changes should be guarded by some `if is_cobalt` construct since it's Chromium code.

File: cobalt/android/BUILD.gn
Comment (oxve) [inline cobalt/android/BUILD.gn:1]: `if is_cobalt` here too

File: cobalt/docker/unittest/Dockerfile [hunk 7]
```diff
@@ -2,6 +2,8 @@ ARG BASE_IMAGE_SHA=sha256:be15f84c44c4ee68b4f187128f0278df1b0f424c04fb5f08b09896
 
 FROM marketplace.gcr.io/google/debian12@${BASE_IMAGE_SHA}
 
+ENV DEBIAN_FRONTEND=noninteractive
```
Comment (oxve) [inline cobalt/docker/unittest/Dockerfile]: Nit/question: Is this line necessary?

File: cobalt/build/archive_test_artifacts.py [hunk 8]
```diff
@@ -141,6 +155,14 @@ def create_archive(
       if os.path.exists(test_targets_json):
         target_deps.add(os.path.join(tar_root, 'test_targets.json'))
 
+      # Add JUnit wrapper scripts if they exist
```
Comment (oxve) [inline cobalt/build/archive_test_artifacts.py]: For gtest targets I've seen the wrapper automatically getting included in the runtime deps. Are you sure these are required to add manually?

File: cobalt/build/archive_test_artifacts.py [hunk 9]
```diff
@@ -23,11 +23,8 @@
 
 # Path prefixes that contain files we don't need to run tests.
 _EXCLUDE_DIRS = [
```
Comment (oxve) [inline cobalt/build/archive_test_artifacts.py]: Can you verify that the android unit test deps archives do not include these? I recall adding them to avoid ballooning them.

File: cobalt/build/archive_test_artifacts.py [hunk 10]
```diff
@@ -111,12 +109,18 @@ def create_archive(
...
       continue
     target_path, target_name = target.split(':')
+    target_path = target_path.lstrip('/')
     # Paths are configured in test.gni:
     # https://github.com/youtube/cobalt/blob/main/testing/test.gni
     if use_android_deps_path:
       deps_file = os.path.join(
           out_dir, 'gen.runtime', target_path,
           f'{target_name}__test_runner_script.runtime_deps')
+      if not os.path.exists(deps_file):
```
Comment (oxve) [inline cobalt/build/archive_test_artifacts.py]: I'm not too fond of this. There was one `if not os.path.exists(deps_file)` before this change. Now there are 5. This must be possible to simplify.

Perhaps add another flag for junit? Or figure out if we can/should add the suffix for junit tests.

File: .github/actions/on_host_tests/action.yaml [hunk 11]
```diff
@@ -27,34 +31,34 @@ runs:
       uses: actions/download-artifact@v5
       with:
         name: ${{ inputs.test_artifacts_key }}
+    - name: Checkout JUnit Dependencies
+      if: ${{ inputs.test_type == 'junit' }}
```
Comment (oxve) [inline .github/actions/on_host_tests/action.yaml]: nit: `${{` are optional in `if:` statements. It makes it slightly harder so we've excluded in in most places.

File: .github/actions/on_host_tests/action.yaml [hunk 12]
```diff
@@ -27,34 +31,34 @@ runs:
       uses: actions/download-artifact@v5
       with:
         name: ${{ inputs.test_artifacts_key }}
+    - name: Checkout JUnit Dependencies
```
Comment (oxve) [inline .github/actions/on_host_tests/action.yaml:34]: We set up the sparse checkout in `main.yaml` so I think I would like to see this happening as soon as possible after that. I'm not feeling 100% sure on this so argue against me if you don't agree.

**Review by oxve — COMMENTED** [review]
The approach looks good. Just some details to iron out.

File: .github/actions/on_host_tests/action.yaml (hunk 4 above)
Comment (arjungm) [reply .github/actions/on_host_tests/action.yaml:39]: I checked the runtime deps file and they're not present: https://paste.googleplex.com/5437630572527616

It seems like it doesn't generate this for the test_runner_script wrappers

File: build/android/test_runner.py
Comment (arjungm) [reply build/android/test_runner.py:1]: yes it's necessary, the problem is because it only applies this argument for Gtest type tests but it's passed by the test wrapper script code when we re-enabled the runtime deps generation

File: cobalt/build/archive_test_artifacts.py (hunk 9 above)
Comment (arjungm) [reply cobalt/build/archive_test_artifacts.py]: it seems like they do increase in size. I tested with base_unittests and it went from from 47 MB to 52 MB

File: cobalt/build/archive_test_artifacts.py (hunk 8 above)
Comment (arjungm) [reply cobalt/build/archive_test_artifacts.py]: this is the example runtime deps file generated for the base_unittests and I see any wrapper scripts in there: https://paste.googleplex.com/5280624050044928

Without these lines for junit tests however it fails to package them and then it doesn't work. I'm thinking that this is once again another special case because of how the robolectric binaries are built in our code base. There seems to be some unfinished migration work on these targets.

Comment (arjungm) [reply cobalt/build/archive_test_artifacts.py]: additionally, for `base_unittests` we see only the `out/android/bin/run_base_unittests` script, whereas for `cobalt_coat_junit_tests` we see both:

```
out/android/bin/run_cobalt_coat_junit_tests
out/android/bin/helper/run_cobalt_coat_junit_tests
```

File: cobalt/build/archive_test_artifacts.py (hunk 9 above)
Comment (arjungm) [reply cobalt/build/archive_test_artifacts.py]: added some logic to ensure the old exclude dirs are applied when it's not a junit test

File: build/android/test_runner.py
Comment (arjungm) [reply build/android/test_runner.py:1]: I'm moved it to the Junit section to miminize the fallout to other test infra

File: .github/actions/on_host_tests/action.yaml (hunk 12 above)
Comment (arjungm) [reply .github/actions/on_host_tests/action.yaml:34]: I'm not sure where I would move this since we need the test_type to be parsed from the test_targets.json prior to being able to gate whether this step runs. I could put it in the main job steps, but then that separates it from where it would be most relevant making for more overhead to track down what earlier step failed. I think having it here is fine given that it's such a specific edge case for Junit tests.

File: .github/actions/on_host_tests/action.yaml (hunk 4 above)
Comment (arjungm) [reply .github/actions/on_host_tests/action.yaml:39]: spoke offline but this could be fixed by having on host test runnersuse the overlay and runner cache with gclient sync steps

File: cobalt/build/archive_test_artifacts.py (hunk 8 above)
Comment (arjungm) [reply cobalt/build/archive_test_artifacts.py]: captured some of this in this refactoring GN for Robolectric bug: 524712602

File: build/android/test_runner.py
Comment (arjungm) [inline build/android/test_runner.py:1]: @johnxwork @Awallky PTAL at this android specific changes

File: build/config/android/rules.gni
Comment (arjungm) [inline build/config/android/rules.gni:1]: @johnxwork @Awallky PTAL at this android specific changes

**Review by johnxwork — APPROVED** [review]
For Android specific test code, it looks like the PR only pipes new variables and should be safe. Fingers crossed everything works

**Review by Awallky — APPROVED** [review]
As long as junit tests run and pass, LGTM
